keps/sig-network/1111-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,681 @@
+subsequently networking related processes).
+
+We can define two classes of services, when the type is `LoadBalancer`.
+
+- A) services which have `externalTrafficPolicy: Cluster`
+- B) services which have `externalTrafficPolicy: Local`
+
+These are the only two classes of services which impact traffic forwarding for a
+load balancer and answer the question: "where traffic should be sent to?".
+
+Load balancers, today, go about figuring out where "traffic should be sent to"
+through the utilization of an HC. The HC is configured differently depending on
+the class of service.
+
+- For services of class B: the HC is configured to probe all nodes in its set
+  on a node port defined to be value of the field `healthCheckNodePort`. The
+  field can be user defined, but when its not: it's automatically generated by
+  Kubernetes. The entity answering to the HC on this port is the service proxy
+  and its answer currently reflects one sole criteria: does the service have an
+  endpoint running on the node in question? If the answer is "yes", then an HTTP
+  status 200 (StatusOK) is returned to the LB, indicating to the LB that it can
+  use the node as a target for traffic forwarding. If the answer is "no", then
+  an HTTP status 503 (StatusServiceUnavailable) is returned, indicating the
+  inverse. Service proxies which implement and support class B type services,
+  must implement the above.
+- For services of class A: it's more of a "free for all". Cloud providers define
+  their own HC, and no one does the same. The following describes what the HC
+  looks like on the three major public cloud providers:
+  - GCP: probes port 10256 (Kube-proxy's healthz port)
+  - AWS: if ELB; probes the first `NodePort` defined on the service spec
+  - Azure: probes all `NodePort` defined on the service spec.
+  
+  All clouds take an approach of trying to ascertain if traffic can be forwarded
+  to the endpoint, which is a completely valid HC for class A services. More is
+  mentioned about this in
+  [Defining what should be "healthy"](#defining-what-should-be-healthy)
+
+To find a solution to the problem of avoiding an ingress outage when a service
+transition from class A <-> B: we will need to align the behavior across service
+classes. That is to say: align service proxies and cloud providers with a
+unified HC which can accurately depict what the load balancer should do for both
+types of services classes, more about this is mentioned in
+["Extend healthCheckNodePort to class A services"](#extend-healthchecknodeport-to-class-a-services)
+
+### Defining what should be "healthy"
+
+When [Defining "healthy"](#defining-healthy), it is important is to keep in mind
+that we are discussing health from the perspective of an
+__application load balancer__.
+
+We could argue that the logic in the KCCM has been flawed all along. Ingress for
+both class A and B services should not care what the state of the node / Kubelet
```

Comment:
this is a strong statement, there is an direct relation between pod and node state, and pods are the backends of the services ... per example, scheduler will start to evict pods on faulty nodes, if you know that you can start draining connections ....
I think that the problem is not the dependency on node state, is the asynchronous relation between the node lifecycle and pod lifecycle ... it has some room to improvement

Reply:
> scheduler will start to evict pods on faulty nodes, if you know that you can start draining connections .... 

But removing the node from the LB set like the CCM does, does not drain. It abruptly kills all connections. kube-proxy which is failing the HC will drain the connections.  

> I think that the problem is not the dependency on node state, is the asynchronous relation between the node lifecycle and pod lifecycle ... it has some room to improvement 

I can re-formulate to this.